src/carma-1-tenth/calc_route_deviation.py

```diff
@@ -0,0 +1,123 @@
+    odom_count = 0
+    odometry = np.zeros((topic_count_dict[odom_topic], 2))
+    for idx in tqdm.tqdm(iterable=range(topic_count_dict[odom_topic] + topic_count_dict[route_topic])):
+        if(reader.has_next()):
+            (topic, data, t_) = reader.read_next()
+            msg_type = type_map[topic]
+            msg_type_full = get_message(msg_type)
+            msg = deserialize_message(data, msg_type_full)
+            if msg_type == "visualization_msgs/msg/MarkerArray":
+                route_graph = msg
+            else:
+                odometry[odom_count] = [-msg.pose.pose.position.y, msg.pose.pose.position.x]
+                odom_count += 1
+    route_coordinates = []
+    for i in range(len(route_graph.markers)):
+        if route_graph.markers[i].type == 2:
+            route_coordinates.append([-route_graph.markers[i].pose.position.y, route_graph.markers[i].pose.position.x])
+    route_coordinates_with_distance = np.zeros((len(route_coordinates), 3))
+    route_coordinates_with_distance[:, 1:] = np.array(route_coordinates)
+    running_distance = 0.0
+    for i in range(len(route_coordinates)):
+        if i > 0:
+            running_distance += np.linalg.norm(route_coordinates_with_distance[i, 1:] - route_coordinates_with_distance[i-1, 1:])
+        route_coordinates_with_distance[i,0] = running_distance
+    x_spline = make_interp_spline(route_coordinates_with_distance[:,0], route_coordinates_with_distance[:,1])
+    y_spline = make_interp_spline(route_coordinates_with_distance[:,0], route_coordinates_with_distance[:,2])
+    samples = np.linspace(route_coordinates_with_distance[0,0], route_coordinates_with_distance[-1,0], 5000)
+    route_x_points = x_spline(samples)
+    route_y_points = y_spline(samples)
+    route_deviations = []
+    odom_movements = []
+    for i in range(1, len(odometry)):
+        closest_point = find_closest_point(np.array([route_x_points, route_y_points]).T, odometry[i])
+        if closest_point is not None and np.linalg.norm(odometry[i]- odometry[i-1]) >= 0.005:
+            route_deviations.append(np.linalg.norm(odometry[i] - closest_point))
+            odom_movements.append(np.linalg.norm(odometry[i]- odometry[i-1]))
+            
+    print("Average Deviation:", np.mean(route_deviations))
+    print("Maximum Deviation:", np.max(route_deviations))
+    plt.plot(range(len(route_deviations)), route_deviations)
+    plt.plot(range(len(odom_movements)), odom_movements)
+
+if __name__=="__main__":
+    import argparse, argcomplete
+    parser = argparse.ArgumentParser(description="Plot steering rate of C1T trucks")
+    parser.add_argument("bag_in", type=str, help="Bag to load")
+    parser.add_argument("--png_out", type=str, help="Output file")
+    argcomplete.autocomplete(parser)
+    args = parser.parse_args()
+    argdict : dict = vars(args)
+    calc_route_deviation(os.path.normpath(os.path.abspath(argdict["bag_in"])))
+    plt.xlabel("Index")
+    plt.ylabel("Deviation from Route (m)")
```

Comment:
Consider "Absolute Deviation from Route" for clarity since we aren't reflecting left-right deviations with signed values

Reply:
Changed to absolution route deviation